subprojects/launcher/src/test/groovy/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutorTest.groovy

```diff
@@ -103,35 +119,152 @@ class ContinuousBuildActionExecutorTest extends ConcurrentSpec {
 
         then:
         1 * delegate.execute(action, buildSessionContext)
-        1 * deploymentRegistry.runningDeployments >> []
-        0 * waiterFactory._
         System.in instanceof DisconnectableInputStream
         System.in.read() == -1
     }
 
-    def "waits for waiter"() {
+    def "can cancel the build"() {
         when:
-        continuousBuild()
-        1 * delegate.execute(action, buildSessionContext) >> {
-            declareInput(file)
+        continuousBuildWithInputs()
+        runningBuild = startBuild()
+
+        then:
+        conditions.eventually {
+            lastLogLine == "Waiting for changes to input files of tasks..."
+        }
+
+        when:
+        cancellationToken.cancel()
+
+        then:
+        buildExits()
+        lastLogLine == "Build cancelled."
+    }
+
+    def "runs a new build on file change"() {
+        when:
+        continuousBuildWithInputs()
+        startBuild()
+
+        then:
+        conditions.eventually {
+            waitingForChanges()
         }
-        executeBuild()
+
+        when:
+        changeListeners.broadcastChange(FileWatcherRegistry.Type.MODIFIED, file.toPath())
 
         then:
-        1 * waiter.wait(_, _) >> {
-            cancellationToken.cancel()
+        conditions.eventually {
+            rebuiltBecauseOfChange()
         }
+
+        when:
+        cancellationToken.cancel()
+        then:
+        buildExits()
```

Comment:
Move this to `cleanup:`? Remove it maybe?

Reply:
Moved to `cleanup:`.